GRAVITY-1030 Add documentation for the authentication of WebRTC In-AppCalling (#1097)
Bandwidth/api-docs: site/global-guides/voice/authentication.mdx

```mermaid
sequenceDiagram
    Application->>In-App Calling AuthServer: GET /api/v1/oauth2/token?scope=voice.webrtc <br> Headers: { Authorization: Basic PEJhc2ljIEF1dGggVXNlcm5hbWU+OjxCYXNpYyBBdXRoIFBhc3N3b3JkPg== }
    In-App Calling AuthServer->>In-App Calling AuthServer: Validate credentials
    In-App Calling AuthServer->>Application: 200 OK <br> {"access_token": "...", "access_token_id": "...", "token_type": "Bearer", "expires_in": 600}
```
### Request
In order to retrieve the token, you will use Basic Authentication `ClientID:ClientSecret`. Below is a sample request:

```sh
curl --request POST \
  --url https://id.voxbone.com/api/v1/oauth2/token?scope=voice.webrtc \
  --header 'Authorization: Basic PEJhc2ljIEF1dGggVXNlcm5hbWU+OjxCYXNpYyBBdXRoIFBhc3N3b3JkPg==' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data grant_type=client_credentials
```
Now that we've shown the request, let's break down the parts of the API call.

> - **url** `https://id.voxbone.com/api/v1/oauth2/token?scope=voice.webrtc` - This is the In-App Calling AuthServers's `token_endpoint`. **IMPORTANT**: There is a query parameter of `scope`. This is used to limit the permissions in the token. When you request a token with the scope parameter that token will **ONLY** be able to access resources that require the specifiied scopes. In our example, we are limiting the token's permissions to only the WebRTC Gateway. If you do not provide a scope, then the token will be minted for **ALL** scopes that are allowed by the credentials when they were created.

Author: I am remove majority of this in favor of our sidebar conversation.

- **Headers**

>   - **Authorization** - This request requires authentication. You **MUST** provide your `ClientID` and `ClientSecret` in a Basic Authorization header. Basic Authentication uses a Base64 encoded string that contains the username and password.

Reviewer: ```suggestion
  - **Authorization** - Authentication is required. You **MUST** provide your `ClientID` and `ClientSecret` through a Base64 encoded Basic Authorization header. The encoded value should take the form `<ClientID>:<ClientSecret>`.
```

Reviewer 2: My original comment changed this wording, but mainly I did not want the term "Basic Authorization header". Although, it's mostly understood what that is, the term "Basic Authentication" is the standard, not "Basic Authorization". The `Basic` is the authentication scheme.

Also, since the usage of `Base64-encoded` is an adjective, it requires a hyphen.

Reviewer 2: ```suggestion
  - **Authorization** - Authentication is required. You **MUST** provide your `ClientID` and `ClientSecret` using a Base64-encoded header with the `Basic` authentication scheme. The header value before encoding should take the form `Basic <ClientID>:<ClientSecret>`.
```
This is what it looks like if @matthewkmartin's and my suggestions are merged.

---
id: inAppAuthGuide
title: Authentication for Global In-App Calling
slug: /voice/guides/inAppCalling/authentication
description: Authentication In-App Calling (WebRTC)
keywords:
  - bandwidth
  - voice
hide_title: false
image: '@site/static/img/bw-icon.svg'
---
## Retrieving an access token

:::caution

> The token should be retrieved from a server running in a secure environment and securely provided to clients. Client-side javascript does not have a mechanism for hiding these credentials so DO NOT place these directly in your client-side code.

Reviewer: ```suggestion
The token should be retrieved from a server running in a secure environment and securely provided to clients. Client-side javascript does not have a mechanism for hiding these credentials so **DO NOT** place these directly in your client-side code.
```
Consistent styling?

Bandwidth accepts no responsibility for the loss of account credentials and any resulting network traffic, fraud, or undesired account access that results from failing to manage account access credentials in a completely secure manner.
:::

> The process to retrieve an access token requires you to have your [WebRTC Credentials](./inAppGuide.mdx#generating-credentials-for-an-access-token) (ie. ClientID and ClientSecret), generated from the [Global Portal](https://login.voxbone.com/login). You will then be able to make a call to our token endpoint with these credentials. This will return to you an `access_token`, `access_token_id`, `token_type`, and `expires_in` (this is the number of seconds before the token will expire).

Reviewer: Should we be specifying the values returned here?  The response section details these values in more detail, so perhaps we should just removed the specifier in the parentheses at the end.

```mermaid
sequenceDiagram
    Application->>BW Global API AuthServer: GET /api/v1/oauth2/token?scope=voice.webrtc <br> Headers: { Authorization: Basic PEJhc2ljIEF1dGggVXNlcm5hbWU+OjxCYXNpYyBBdXRoIFBhc3N3b3JkPg== }
    BW Global API AuthServer->>BW Global API AuthServer: Validate credentials
    BW Global API AuthServer->>Application: 200 OK <br> {"access_token": "...", "access_token_id": "...", "token_type": "Bearer", "expires_in": 600}
    Application->>WebRTC Gateway: GET /someresource <br> Headers: {Authorization: <token_type> <access_token>}
```
### Request
In order to retrieve the token you will use Basic Auth `ClientID:ClientSecret`. Below is a sample request

```sh
curl --request POST \
  --url https://<voxbones_token_providerd_url>/api/v1/oauth2/token?scope=voice.webrtc \
  --header 'Authorization: Basic PEJhc2ljIEF1dGggVXNlcm5hbWU+OjxCYXNpYyBBdXRoIFBhc3N3b3JkPg==' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data grant_type=client_credentials
```

> Now that we've shown the request let's break down the parts of the call.

Reviewer 2: ```suggestion
Now that we've shown the request let's break down the parts of the API call.
```
Because we are telecom, we have voice calls and I usually like to make the distinction of an API call because it's not always so. "Call" can mean a voice call.

> Now that we've shown the request let's break down the parts of the API call.

Reviewer: ```suggestion
Now that we've shown the request, let's break down the parts of the API call.
```

Reviewer 2: Moving towards the Shatner Comma style? 😆  I agree. Comma is correct.

> - **url** `https://<voxbones_token_provider_url>/api/v1/oauth2/token?scope=voice.webrtc` - This is the BW Global API AuthProvider's `token_endpoint`. You may notice there is a query parameter of `scope` this is used to limit the permissions in the token. When you request a token with the scope parameter that token will **ONLY** be able to talk to resources that require those particular scopes which were requested. In our example, we are limiting the token to only being able to talk to the WebRTC Gateway. If you do not provide a scope then the token will be minted for ALL the scopes that the credentials are allowed to have as defined during creation.

Reviewer 2: ```suggestion
- **url** `https://<voxbones_token_provider_url>/api/v1/oauth2/token?scope=voice.webrtc` - This is the BW Global API AuthProvider's `token_endpoint`. **IMPORTANT:** There is a query parameter of `scope`. This is used to limit the permissions in the token. When you request a token with the scope parameter, that token will **ONLY** be able to access resources that require the specified scopes. In our example, we are limiting the token's permissions to only the WebRTC Gateway. If you do not provide a scope, then the token will be minted for ALL scopes that are allowed by the credentials when they were created.
```

Reviewer 2: - The `<voxbone url>` will need replacing.
- You used the term `AuthProvider` but you were saying `AuthServer` prior to here. If you agree with my previous terminology of `In-App Calling Auth Server`, then it won't matter.

> - **url** `https://id.voxbone.com/api/v1/oauth2/token?scope=voice.webrtc` - This is the In-App Calling AuthServers's `token_endpoint`. **IMPORTANT** There is a query parameter of `scope`. This is used to limit the permissions in the token. When you request a token with the scope parameter that token will **ONLY** be able to access resources that require the specifiied scopes. In our example, we are limiting the token's permissions to only the WebRTC Gateway. If you do not provide a scope, then the token will be minted for ALL scopes that are allowed by the credentials when they were created.

Reviewer: ```suggestion
- **url** `https://id.voxbone.com/api/v1/oauth2/token?scope=voice.webrtc` - This is the In-App Calling AuthServers's `token_endpoint`. **IMPORTANT**: There is a query parameter of `scope`. This is used to limit the permissions in the token. When you request a token with the scope parameter that token will **ONLY** be able to access resources that require the specifiied scopes. In our example, we are limiting the token's permissions to only the WebRTC Gateway. If you do not provide a scope, then the token will be minted for **ALL** scopes that are allowed by the credentials when they were created.
```
For consistent styling?

- **Headers**

>   - **Authorization** - This request requires the request to be authenticated. You MUST provide your `ClientID` and `ClientSecret` in a Basic Authorization header. Basic Authentication is a Base64 encoded string that contains the username and password.

Reviewer 2: Slightly different wording was used based on @matthewkmartin's comments.

>   - **Content-Type** - For this request we **MUST** to use `application/x-www-form-urlencoded`. This will URL encode (aka percent-encode) our parameters making them into `key=value` pairs which will be separated by an ampersand `&`.

Reviewer: ```suggestion
  - **Content-Type** - For this request you **MUST** use `application/x-www-form-urlencoded`. This will URL encode (aka percent-encode) the parameters making them into `key=value` pairs which will be separated by an ampersand `&`.
```

> - **Body** - There is only one piece of data that must be transmitted in the form and that is the `grant_type` currently we only support `client_credentials` OAuth2 `grant_type`.

Reviewer 2: ```suggestion
- **Body** - There is only one piece of data that must be transmitted in the form and that is the `grant_type`. Currently we only support `client_credentials` OAuth2 `grant_type`.
```

### Response

> Am example response is:

Reviewer 2: ```suggestion
An example response is:
```
I would have just fixed the "Am" typo.

```
{
	"access_token": "eyJraWQiOiJzZ25tLTk1ZTE0YTFiLTRiODItNDM5Zi1hYjhmLTYwMmI5MzYwMjZkZiIsImFsZyI6IlJTMjU2In0..kl3HYFQIcP4xZd78wUkdXxntxBu1d7b5ZCt99qVObC-gUZd5KSq2J7Q5rDb2LjP1_WVndcLQSqCoq3Anvp03hJWM6mamZL3dWHxjGLwkrIlzmNx9ZFGhTGIEsYLyaQqy8MonWYw2mJ4Z0APEfTVbTBHNIGrm_9GT6rIkdOwQFM-XgH1Tau4JbpFJun3n6o15WTgjzdEj9fIwd385CPwL-pAmOiozbYWUEzgqkXjSGHI11hiLDu-tv_8Ds06Cx4iCnL1F6_dFrnpD3CF0i6JJYVrvLmi6vgzoxp9YEIRBwaOZTSuYuYt03SQfbMZy8L2Z71sRKLLGt3TrxgtwyjefpQ",
	"access_token_id": "a5xA3xMKEggGwvpSLtk2lRb",
	"token_type": "Bearer",
	"expires_in": 600
}
```

Let's see what each item in the response object is meant for:
- `access_token`: This is the JSON Web Token (JWT) that will be passed in the Authorization header of the requests to the In-App Calling API.
- `access_token_id`: This will be the ID of tha access token to prevent tokens from being re-used. It will be the same as the jti claim within the access token.
- `token_type`: This is used to identify the type of token and also the prefix to be used in the Authorization header.

> - `expires_in`: This is how long in seconds the token will be valid. In the example above the value is `600`, this means the token will be valid for 600 seconds or 10 mins. You **SHOULD** use the token until 10-30 seconds before expiration at which point you should request a new token in the same manner as described above.

Reviewer 2: ```suggestion
- `expires_in`: This is the length of time, in seconds, before a token expires. In the example above, the value is `600`. This means the token will be valid for 600 seconds (10 minutes) before it expires. You **SHOULD** use the token until 10-30 seconds before expiration, at which point you must request a new token in the same manner as described above.
```

### Using the Access Token

> Once you have the token you can now perform authenticated requests to the WebRTC Gateway, all you need to do is add an `Authorization` header to the request with a value of `<token_type> <access_token>`.

Reviewer 2: Section has been replaced with a link to WebRTC SDK.

Using the example response above a request would look similar to:
```sh
curl --request GET \

>   --url https://<webrtc_gateway>/some/protected/path \

Reviewer 2: Not sure what or how we described the `webrtc_gateway` value.

Author: I am not sure the of the URL perhaps we can get this from Pritom or Preo.

Reviewer 2: I think it's `sbc.webrtc-app.bandwidth.com`.
@vikash-band What is the correct URL?

Reviewer 3: sbc.webrtc-app.bandwidth.com is correct url

> Using the example response above a request would look similar to:

Reviewer: ```suggestion
Using the values from the example response listed above, a sample request would look like:
```
